Dsc/pras2psy updates

scdhulipala opened the pull request:
Closes https://github.com/NREL-Sienna/SiennaPRASInterface.jl/issues/54

This uses the RATemplate and DeviceModel concepts we now have.

Commit:
Updates to PRAS2PSY function signatures

Commit:
Updates to assess to generate required results to generate outage profile

Commit:
Final updates to PRAS2PowerSystems.jl and add helper functions

Commit:
Add test and run formatter

josephmckinsey commented on src/PRAS2PowerSystems.jl:
```diff
@@ -1,240 +1,265 @@
+  - PSY System with PSY.TimeSeriesForcedOutage for all components for which asset status is available
+"""
+function generate_outage_profile!(
+    sys::PSY.System,
```
This one should call the other `generate_outage_profile!` with the template and all that.

scdhulipala replied:
Done.

josephmckinsey commented on src/PRAS2PowerSystems.jl:
```diff
@@ -1,240 +1,265 @@
-    end
+    - PSY System with PSY.TimeSeriesForcedOutage for all components for which asset status is available
+"""
+function generate_outage_profile!(sys::PSY.System, method::PRASCore.SequentialMonteCarlo)
```
This one should also call the `generate_outage_profile!` with RATemplate.

scdhulipala replied:
DOne.

josephmckinsey commented on src/PRAS2PowerSystems.jl:
```diff
@@ -1,240 +1,265 @@
+
+ - PSY System with PSY.TimeSeriesForcedOutage for all components for which asset status is available
+"""
+function generate_outage_profile!(
```
Should we have a function that takes a PRAS system still?

jd-lara replied:
Probably not. The PRAS system should be hidden from the user.

scdhulipala replied:
Also, you finally need a PSY System to add this time series to. This will make things messy and we have no control of mapping from the PRAS System and PSY System the user is passing.

josephmckinsey commented on src/PRAS2PowerSystems.jl:
```diff
@@ -1,240 +1,265 @@
+    ts_timestamps = TimeSeries.timestamp(first(all_ts).data)
+
+    shortfall_samp_idx =
+        findfirst(name.(typeof.(results)) .== PRASCore.Results.ShortfallSamplesResult)
```
Is there really not a way to make this not dynamic?

jd-lara replied:
this code `findfirst(name.(typeof.(results)) .== PRASCore.Results.ShortfallSamplesResult)` smells hacky....what are you trying to do here?

scdhulipala replied:
The PRAS.Assess() call returns a tuple of results in whatever order ResultSpecs are passed. We can't say for sure this is the index of ResultType. I need certain ResultType to do certain things. Here I'm trying to find the index of ResultType if that makes sense.

scdhulipala replied:
I'm not doing this anymore. I just pushed a fix which should make this more _elegant_ and less hacky I think.

scdhulipala replied:
The change I pushed eliminates this. I am now defining a struct to parse the PRASResult tuple.

josephmckinsey commented on src/PRAS2PowerSystems.jl:
```diff
@@ -1,240 +1,265 @@
+    shortfall_samples = results[shortfall_samp_idx]
+
+    sample_idx = sortperm(shortfall_samples[], rev=true)[1]
+    resultsspecs = get_outage_pras_resultspec(template)
```
In our cases, this should be exactly the resultspecs we compute with earlier, right?

scdhulipala replied:
DOne. I'm passing this to add_asset_status! now.

josephmckinsey commented on src/PRAS2PowerSystems.jl:
```diff
@@ -1,240 +1,265 @@
+        findfirst(name.(typeof.(results)) .== PRASCore.Results.ShortfallSamplesResult)
+    shortfall_samples = results[shortfall_samp_idx]
+
+    sample_idx = sortperm(shortfall_samples[], rev=true)[1]
```
Is there a way to get the maximum index without sorting everything?

scdhulipala replied:
Done

josephmckinsey commented on src/PRAS2PowerSystems.jl:
```diff
@@ -1,240 +1,265 @@
+
+            availability_data = TimeSeries.TimeArray(
+                ts_timestamps,
+                getindex.(gens_avail[gen.name, :], sample_idx),
```
This is where we actually use the relationship between names in PRAS and Sienna, if I understand it correctly?

scdhulipala replied:
Yes, this is correct.

josephmckinsey commented on src/PRAS2PowerSystems.jl:
```diff
@@ -1,240 +1,265 @@
+"""
+Get name of Type to strip out parametes for easy comparison
+"""
+name(::Type{T}) where {T} = (isempty(T.parameters) ? T : T.name.wrapper)
```
We should rename this to something other than `name`.

scdhulipala replied:
I renamed it to _type_name_.

scdhulipala replied:
I removed this in my latest commit because we don't need this anymore.

josephmckinsey commented on src/PRAS2PowerSystems.jl:
```diff
@@ -1,240 +1,265 @@
+    results::T,
+    template::RATemplate,
+) where {T <: Tuple{Vararg{PRASCore.Results.Result}}}
+    @info ""
```
This `@info` shouldn't be there.

josephmckinsey commented on src/PRAS2PowerSystems.jl:
```diff
@@ -1,240 +1,265 @@
+    @info ""
+    # Time series timestamps
+    filter_func = x -> (typeof(x) <: PSY.StaticTimeSeries)
+    all_ts = PSY.get_time_series_multiple(sys, filter_func)
```
This is the only place where `filter_func` is used.

jd-lara replied:
This is pretty bad. We should be using the `has_time_series` and `get_time_series` correctly. @daniel-thom can you review this line. The code here seems odd to me and I am not sure it is using the time series management features correctly.

scdhulipala replied:
I'm not sure I totally understand where you want me to use has_time_Series and get_time_series. I will follow up with @daniel-thom about what I'm doing here.

scdhulipala replied:
> This is the only place where `filter_func` is used.

I'm not defining filter_func anymore.

josephmckinsey reviewed (changes requested):
I only reviewed the changes from the last 4 commits.

Commit:
Changes to address @josephmckinsey comments

Commit:
Changes to address @jd-lara comments

Commit:
Add struct to parse PRASResult tuple

scdhulipala commented:
I think I addressed all your comments. The only outstanding issue is addressing @jd-lara  comment on how I was getting the timestamps. I can address these once PSY is tageed. I will do this on a separate PR along with addressing the lumping issue.

Commit:
Remove passing resultsspecs to add_asset_status and add new test

Commit:
Simplify the logic of PRAS2PSY
